fix(ivy):track changes across failed builds

alxhub opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -12,49 +12,155 @@ import {DependencyTracker} from '../../partial_evaluator';
+        // It's a .ts file, so track it as a change.
+        state.changedTsPaths.add(newFile.fileName);
+      } else {
+        // It's a .d.ts file. Currently the compiler does not do a great job of tracking
```
I would move this branch to be the very first in the loop

alxhub replied:
The loop has early exits, so this would break the logic. I think the way it's organized is as readable as possible currently.

JoostK replied:
Oh, I realized that during my initial pass, and then completely missed it when typing this comment. d'oh!

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -12,49 +12,155 @@ import {DependencyTracker} from '../../partial_evaluator';
+  static fresh(program: ts.Program): IncrementalDriver {
+    // Initialize the set of files which need to be emitted to the set of all TS files in the
+    // program.
+    const tsFiles = program.getSourceFiles().filter(sf => !sf.isDeclarationFile);
```
```suggestion
    const tsFiles = tsOnlyFiles(program.getSourceFiles());
```

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -12,49 +12,155 @@ import {DependencyTracker} from '../../partial_evaluator';
+      const fileHasNoChangedDependencies =
+          !this.getFileDependencies(sf).some(dep => state.changedTsPaths.has(dep.fileName));
+
+      if (isFileItselfUnchanged && fileHasNoChangedDependencies && fileHasNoChangedResources) {
```
I get what you're doing here, however I'd find it better readable if all conditions were not inverted and the `pendingEmit.add(sf.fileName)` was moved within the body here.

JoostK replied:
Also, consider replacing the variables with methods/functions so that we benefit from short circuiting here.

alxhub replied:
Done! I just inlined everything since it's still very readable (and we benefit from short-circuiting as you mentioned).

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -12,49 +12,155 @@ import {DependencyTracker} from '../../partial_evaluator';
 
+  recordSuccessfulEmit(sf: ts.SourceFile): void { this.state.pendingEmit.delete(sf.fileName); }
+
+  safeToSkip(sf: ts.SourceFile): boolean { return !this.state.pendingEmit.has(sf.fileName); }
```
We should rename this to reflect that this is only targeting emits.

alxhub replied:
Done!

JoostK commented on packages/compiler-cli/src/ngtsc/incremental/src/state.ts:
```diff
@@ -87,12 +193,14 @@ export class IncrementalState implements DependencyTracker, ResourceDependencyRe
     return Array.from(resourceDeps.keys())
-        .some(resourcePath => this.modifiedResourceFiles !.has(resourcePath));
+        .some(
+            resourcePath => this.state.kind === BuildStateKind.Pending &&
```
`this.state.kind === BuildStateKind.Pending` is an invariant here, so it shouldn't have to be within the predicate function. I don't think this is intentional?

alxhub replied:
Sadly, not. Even if it were something known to the outer function, inside the arrow function TS resets all the narrowed types (because there's no indication _when_ the arrow function will be called). So the assertion is a necessary one.

JoostK commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -334,6 +343,19 @@ export class NgtscProgram implements api.Program {
+                continue;
+              }
+
+              const originalSf = ts.getOriginalNode(writtenSf) as ts.SourceFile;
```
I am inclined to move this remapping into the incremental driver. Also, this is probably redundant as `recordSuccessfulEmit` only captures the file name, which I presume won't change for transformed source files?

alxhub replied:
Good point. Removed.

alxhub committed:
fix(ivy): track changes across failed builds

Previously, our incremental build system kept track of the changes between
the current compilation and the previous one, and used its knowledge of
inter-file dependencies to evaluate the impact of each change and emit the
right set of output files.

However, a problem arose if the compiler was not able to extract a
dependency graph successfully. This typically happens if the input program
contains errors. In this case the Angular analysis part of compilation is
never executed.

If a file changed in one of these failed builds, in the next build it
appears unchanged. This means that the compiler "forgets" to emit it!

To fix this problem, the compiler needs to know the set of changes made
_since the last successful build_, not simply since the last invocation.

This commit changes the incremental state system to much more explicitly
pass information from the previous to the next compilation, and in the
process to keep track of changes across multiple failed builds, until the
program can be analyzed successfully and the results of those changes
incorporated into the emit plan.

Fixes #32214

alxhub commented:
[Presubmit](http://test/OCL:281845320:BASE:281845349:1574378626963:198d6e21)
Ivy Presubmit: not needed; incremental build doesn't affect g3

JoostK commented on packages/compiler-cli/test/ngtsc/incremental_error_spec.ts:
```diff
@@ -288,5 +395,134 @@ runInEachFileSystem(() => {
+        env.driveMain();
+        env.flushWrittenFileTracking();
+
+        // Update ACmp to have a different selector, which doesn't match BCmp.
```
```suggestion
        // Update ACmp to have a different selector, which isn't matched in BCmp's template.
```

JoostK commented on packages/compiler-cli/test/ngtsc/incremental_error_spec.ts:
```diff
@@ -288,5 +395,134 @@ runInEachFileSystem(() => {
+        env.driveMain();
+
+        // If the compiler behaves correctly, it should remember that 'a.html' was updated before,
+        // and should regenerate a.ts along with the whole module system.
```
"along with the whole module system" sounds wrong to me here. In fact, the comment below rightfully mentions that `b.ts` doesn't need to be reemitted for a resource dependency. As such, this comment seems inaccurate to me.

alxhub replied:
👍

JoostK reviewed (approved):
LGTM, thanks for writing the descriptive comments!

IgorMinar reviewed (approved):
great docs updates!

I'm adding lgtm to deal with codeowners issues until Joost is a member of angular/fw-compiler

IgorMinar commented:
merge-assistance: global approval
